luxonis_train/callbacks/gradcam_visializer.py

```diff
@@ -0,0 +1,205 @@
+        """
+
+        self.model = ModelWrapper(pl_module, self.task)
+
+    def on_validation_batch_end(
+        self,
+        trainer: pl.Trainer,
+        pl_module: pl.LightningModule,
+        outputs: STEP_OUTPUT,
+        batch: Any,
+        batch_idx: int,
+    ) -> None:
+        """At the end of first n batches, visualize the gradients using
+        Grad-CAM.
+
+        @type trainer: pl.Trainer
+        @param trainer: The PyTorch Lightning trainer.
+        @type pl_module: pl.LightningModule
+        @param pl_module: The PyTorch Lightning module.
+        @type outputs: STEP_OUTPUT
+        @param outputs: The output of the model.
+        @type batch: Any
+        @param batch: The input batch.
+        @type batch_idx: int
+        @param batch_idx: The index of the batch.
+        """
+
+        if batch_idx < self.log_n_batches:
+            self.visualize_gradients(trainer, pl_module, batch, batch_idx)
+
+    def visualize_gradients(
+        self,
+        trainer: pl.Trainer,
+        pl_module: pl.LightningModule,
+        batch: Any,
+        batch_idx: int,
+    ) -> None:
+        """Visualizes the gradients using Grad-CAM.
+
+        @type trainer: pl.Trainer
+        @param trainer: The PyTorch Lightning trainer.
+        @type pl_module: pl.LightningModule
+        @param pl_module: The PyTorch Lightning module.
+        @type batch: Any
+        @param batch: The input batch.
+        @type batch_idx: int
+        @param batch_idx: The index of the batch.
+        """
+
+        target_layers = [m[1] for m in pl_module.named_modules()][
+            self.target_layer : self.target_layer + 1
+        ]
+        self.gradcam = HiResCAM(self.model, target_layers)
+
+        x, y = batch
+        model_input = x["image"]
+
+        if self.task == "segmentation":
+            output = self.model(model_input)
+            normalized_masks = torch.nn.functional.softmax(output, dim=1).cpu()
+            mask = normalized_masks.argmax(axis=1).detach().cpu().numpy()
+            mask_float = np.float32(mask == self.class_idx)
+            targets = [
+                SemanticSegmentationTarget(self.class_idx, mask_float[i])
+                for i in range(mask_float.shape[0])
+            ]
+        else:
+            targets = [
+                ClassifierOutputTarget(self.class_idx)
+            ] * model_input.size(0)
+
+        with torch.enable_grad():
+            grayscale_cams = self.gradcam(
+                input_tensor=model_input, targets=targets
+            )
+
+        images = get_denormalized_images(pl_module.cfg, x).cpu().numpy()
+        for zip_idx, (image, grayscale_cam) in enumerate(
+            zip(images, grayscale_cams)
+        ):
+            image = image / 255.0
+            image = image.transpose(1, 2, 0)
+            visualization = show_cam_on_image(
+                image, grayscale_cam, use_rgb=True
+            )
+
+            fig, ax = plt.subplots(1, 1)
```

Comment:
Nitpick: Can we replace the use of matplotlib with cv2 so we don't need to create figures. There is `cv2.applyColorMap(im, cv2.COLORMAP_JET)` function which might suffice?